core/src/main/java/hudson/model/labels/LabelExpression.java

```diff
@@ -210,4 +225,158 @@ public LabelOperatorPrecedence precedence() {
             return LabelOperatorPrecedence.IMPLIES;
         }
     }
+
+    //region Auto-Completion and Validation
+
+    /**
+     * Utility class for taking the current input value and computing a list of potential terms to match against the
+     * list of defined labels.
+     */
+    static class AutoCompleteSeeder {
+        private String source;
+
+        AutoCompleteSeeder(String source) {
+            this.source = source;
+        }
+
+        List<String> getSeeds() {
+            ArrayList<String> terms = new ArrayList<>();
+            boolean trailingQuote = source.endsWith("\"");
+            boolean leadingQuote = source.startsWith("\"");
+            boolean trailingSpace = source.endsWith(" ");
+
+            if (trailingQuote || (trailingSpace && !leadingQuote)) {
+                terms.add("");
+            } else {
+                if (leadingQuote) {
+                    int quote = source.lastIndexOf('"');
+                    if (quote == 0) {
+                        terms.add(source.substring(1));
+                    } else {
+                        terms.add("");
+                    }
+                } else {
+                    int space = source.lastIndexOf(' ');
+                    if (space > -1) {
+                        terms.add(source.substring(space+1));
+                    } else {
+                        terms.add(source);
+                    }
+                }
+            }
+
+            return terms;
+        }
+    }
+
+    /**
+     * Plugins may want to contribute additional restrictions on the use of specific labels for specific jobs.
+     * This extension point allows such restrictions.
+     *
+     * @since TODO
+     */
+    public static abstract class LabelValidator implements ExtensionPoint {
+
+        /**
+         * Validates the use of a label within a job context.
+         *
+         * @param job   The job that wants to restrict itself to the specified label.
+         * @param label The label that the job wants to restrict itself to.
+         * @return The validation result.
+         */
+        @NonNull
+        public abstract FormValidation check(@NonNull Job<?, ?> job, @NonNull Label label);
```

Comment:
Why not make `job` be `@CheckForNull` in case a validator can work without a job context?

Or relax the context type to `Item`?

Reply:
I think the validator concept was added specifically for project-based validation (the main validation code already handles pure label expression validation). It was certainly only ever invoked when a project was specified.

But given that this is new API specifically to relax the Project-oriented existing API, I have no issue using Item instead. I lack the familiarity with the codebase to know what additional cases that would enable though.

Reply:
> I lack the familiarity with the codebase to know what additional cases that would enable though.

Mainly validation on `Folder`s, in case you have something that offers certain labels only in some subfolders. Something like https://github.com/jenkinsci/kubernetes-plugin/blob/d9cf6a810fce807fe35cd56ebe17ffb343c702f4/src/main/resources/org/csanchez/jenkins/plugins/kubernetes/KubernetesFolderProperty/config.jelly#L5 perhaps. (CloudBees CI also has a proprietary example I think.)

Reply:
Done - `Item` is now used instead of `Job`.